brvs-rules/src/main/java/iroha/validation/rules/impl/billing/BillingRule.java

```diff
@@ -0,0 +1,251 @@
+import java.net.URL;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+public class BillingRule implements Rule {
+
+  private static final String QUEUE_NAME = "brvs_billing_updates";
+  private static final String TRANSFER_BILLING_ACCOUNT_NAME = "transfer_billing";
+  private static final String CUSTODY_BILLING_ACCOUNT_NAME = "custody_billing";
+  private static final String ACCOUNT_CREATION_BILLING_ACCOUNT_NAME = "account_creation_billing";
+  private static final String EXCHANGE_BILLING_ACCOUNT_NAME = "exchange_billing";
+  private static final String WITHDRAWAL_BILLING_ACCOUNT_NAME = "withdrawal_billing";
+  private static final Map<String, BillingTypeEnum> feeAccounts = new HashMap<String, BillingTypeEnum>() {{
+    put(TRANSFER_BILLING_ACCOUNT_NAME, BillingTypeEnum.TRANSFER);
+    put(CUSTODY_BILLING_ACCOUNT_NAME, BillingTypeEnum.CUSTODY);
+    put(ACCOUNT_CREATION_BILLING_ACCOUNT_NAME, BillingTypeEnum.ACCOUNT_CREATION);
+    put(EXCHANGE_BILLING_ACCOUNT_NAME, BillingTypeEnum.EXCHANGE);
+    put(WITHDRAWAL_BILLING_ACCOUNT_NAME, BillingTypeEnum.WITHDRAWAL);
+  }};
+  private static final JsonParser jsonParser = new JsonParser();
+  private static final TransferAsset FEE_NOT_NEEDED = TransferAsset.getDefaultInstance();
+  private static final BigDecimal NO_FEE = BigDecimal.ZERO;
+
+  private boolean isRunning = false;
+  private final String getBillingURL;
+  private final String rmqHost;
+  private final int rmqPort;
+  private final String rmqExchange;
+  private final String rmqRoutingKey;
+  private final Gson gson = new Gson();
+  private final Set<BillingInfo> cache = new HashSet<>();
+
+  public BillingRule(String getBillingURL,
+      String rmqHost,
+      int rmqPort,
+      String rmqExchange,
+      String rmqRoutingKey) throws IOException {
+
+    if (Strings.isNullOrEmpty(getBillingURL)) {
+      throw new IllegalArgumentException("Billing URL must not be neither null nor empty");
+    }
+    if (Strings.isNullOrEmpty(rmqHost)) {
+      throw new IllegalArgumentException("RMQ host must not be neither null nor empty");
+    }
+    if (rmqPort < 1 || rmqPort > 65535) {
+      throw new IllegalArgumentException("RMQ port must be valid");
+    }
+    if (Strings.isNullOrEmpty(rmqExchange)) {
+      throw new IllegalArgumentException("RMQ exchange must not be neither null nor empty");
+    }
+    if (Strings.isNullOrEmpty(rmqRoutingKey)) {
+      throw new IllegalArgumentException("RMQ routing key must not be neither null nor empty");
+    }
+
+    this.getBillingURL = getBillingURL;
+    this.rmqHost = rmqHost;
+    this.rmqPort = rmqPort;
+    this.rmqExchange = rmqExchange;
+    this.rmqRoutingKey = rmqRoutingKey;
+    runCacheUpdater();
+  }
+
+  private void runCacheUpdater() throws IOException {
```

Comment:
Is this method called from different threads?

Reply:
No